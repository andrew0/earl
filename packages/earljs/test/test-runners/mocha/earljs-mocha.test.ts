import { expect } from 'chai'
// @ts-expect-error missing typings, but they match child_process.spawn
import _spawn from 'cross-spawn-with-kill'

const spawn = _spawn as typeof import('child_process').spawn

describe('earljs/mocha end-to-end tests', () => {
  it('works in watch mode', async () => {
    // There might be a way to handle it in a more elegant manner...
    const PASSING_TESTS = 1
    const FAILING_TESTS = 1

    const results = await runMochaWatch()

    expect({ passing: results.passing, failing: results.failing }).to.deep.equal(
      {
        passing: PASSING_TESTS,
        failing: FAILING_TESTS,
      },
      `Expected to pass ${PASSING_TESTS} instead of ${results.passing} and fail ${FAILING_TESTS} instead of ${results.failing}.\n` +
        `\nSTDOUT:\n\`${results.stdout}\`` +
        `\nSTDERR:\n\`${results.stderr}\`\n\n`,
    )
  })
})

function runMochaWatch() {
  return new Promise<{ passing: number; failing: number; stdout: string; stderr: string }>((resolve) => {
    const child = spawn('mocha', ['--config', './mocha.config.js', '--watch'], {
      env: process.env,
      cwd: __dirname,
    }).on('error', (err) => {
      throw err
    })

    const result = { passing: NaN, failing: NaN, stdout: '', stderr: '' }

    child.stderr.on('data', (data) => {
      const str = String(data)
      result.stderr += str

      if (str.includes('[mocha] waiting for changes...')) {
        child.kill('SIGINT')
        resolve(result)
      }
    })
    child.stdout.on('data', (data) => {
      const str = String(data)
      result.stdout += str

      const passing = str.match(/(\d+) passing/)
      if (passing) {
        result.passing = parseInt(passing[1])
      }
      const failing = str.match(/(\d+) failing/)
      if (failing) {
        result.failing = parseInt(failing[1])
      }
    })
  })
}
