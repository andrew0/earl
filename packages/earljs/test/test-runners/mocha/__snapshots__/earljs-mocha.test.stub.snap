// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`earljs/mocha failing test 1`] = `0.5506113757632223`;

exports[`earljs/mocha failing test 2`] = `0.6554106565389253`;

exports[`earljs/mocha failing test 3`] = `0.1794920777450828`;

exports[`earljs/mocha failing test 4`] = `0.18139369523734383`;

exports[`earljs/mocha failing test 5`] = `0.5575046137232706`;

exports[`earljs/mocha failing test 6`] = `0.9517831336482709`;

exports[`earljs/mocha failing test 7`] = `0.9007501305802494`;

exports[`earljs/mocha failing test 8`] = `0.9570616377620222`;

exports[`earljs/mocha failing test 9`] = `0.8602380107100129`;

exports[`earljs/mocha failing test 10`] = `0.4421442115508365`;

exports[`earljs/mocha failing test 11`] = `0.6803489827286553`;

exports[`earljs/mocha failing test 12`] = `0.7169182110712202`;

exports[`earljs/mocha failing test 13`] = `0.6620858316086671`;

exports[`earljs/mocha successful test 1`] = `2`;

exports[`earljs/mocha successful test 2`] = `2`;

exports[`earljs/mocha successful test 3`] = `2`;

exports[`earljs/mocha successful test 4`] = `2`;

exports[`earljs/mocha successful test 5`] = `2`;

exports[`earljs/mocha successful test 6`] = `2`;

exports[`earljs/mocha successful test 7`] = `2`;

exports[`earljs/mocha successful test 8`] = `2`;

exports[`earljs/mocha successful test 9`] = `2`;

exports[`earljs/mocha successful test 10`] = `2`;

exports[`earljs/mocha successful test 11`] = `2`;

exports[`earljs/mocha successful test 12`] = `2`;

exports[`earljs/mocha successful test 13`] = `2`;
